File: sdk/go/mics-hook-sdk/micshooksdk/mq.go
package micshooksdk

type MqVerifyResult struct {
	Ok     bool
	Reason string
	Event  MqEvent
}

func EventTopicName(tenantID string) string {
	return "im-mics-" + tenantID + "-event"
}

func DlqTopicName(tenantID string) string {
	return "im-mics-" + tenantID + "-event-dlq"
}

func DecodeMqEvent(b []byte) (MqEvent, error) {
	return decodeMqEvent(b)
}

func EncodeMqEvent(evt MqEvent) []byte {
	return encodeMqEvent(evt)
}

func DecodeAndVerifyMqEvent(b []byte, tenantSecret string, requireSign bool) (MqVerifyResult, error) {
	evt, err := decodeMqEvent(b)
	if err != nil {
		return MqVerifyResult{}, err
	}

	if !requireSign && evt.Sign == "" {
		return MqVerifyResult{Ok: true, Reason: "", Event: evt}, nil
	}
	if evt.Sign == "" {
		return MqVerifyResult{Ok: false, Reason: "invalid sign", Event: evt}, nil
	}

	payloadForSign := encodeMqEvent(clearMqSign(evt))
	if !VerifyMqEventSignBase64(tenantSecret, payloadForSign, evt.Sign) {
		return MqVerifyResult{Ok: false, Reason: "invalid sign", Event: evt}, nil
	}

	return MqVerifyResult{Ok: true, Reason: "", Event: evt}, nil
}

func clearMqSign(evt MqEvent) MqEvent {
	evt.Sign = ""
	return evt
}

File: sdk/node/mics-hook-sdk/src/mq.ts
import { verifyMqEventSign } from "./signing.js";
import { MqEventCodec, type MqEvent } from "./proto/mics_hook.js";

export function eventTopicName(tenantId: string): string {
  return `im-mics-${tenantId}-event`;
}

export function dlqTopicName(tenantId: string): string {
  return `im-mics-${tenantId}-event-dlq`;
}

export function decodeMqEvent(bytes: Uint8Array): MqEvent {
  return MqEventCodec.decode(bytes);
}

export function isValidMqEventSign(tenantSecret: string, evt: MqEvent): boolean {
  if (!evt.sign) return false;
  return verifyMqEventSign(tenantSecret, evt, evt.sign);
}

export function decodeAndVerifyMqEvent(
  bytes: Uint8Array,
  options: { tenantSecret: string; requireSign: boolean }
): { ok: true; event: MqEvent; reason: "" } | { ok: false; event: MqEvent; reason: string } {
  const event = decodeMqEvent(bytes);

  if (!options.requireSign && !event.sign) {
    return { ok: true, event, reason: "" };
  }

  if (!event.sign) {
    return { ok: false, event, reason: "invalid sign" };
  }

  return isValidMqEventSign(options.tenantSecret, event)
    ? { ok: true, event, reason: "" }
    : { ok: false, event, reason: "invalid sign" };
}
